### PR TITLE
Fixes 2443: API support for list / search package groups

### PR DESCRIPTION
**Pull request by xbhouse**
## Summary

This adds:

- `repositories/{uuid}/package_groups` endpoint for listing, returns ID, Name, Description, and Package List metadata
- `package_groups/names` endpoint for searching, returns Name, Description, and Package List of all groups found across the repositories requested
- database support to store package group metadata
- groups are de-duplicated based on combination of name/ID/package list
- search API returns full list of packages that are associated with the search term and repositories

## Testing steps

Endpoint: `$BASE_URL/package_groups/names`

Example request: `{
  "urls": [
    "https://dl.fedoraproject.org/pub/epel/8/Everything/x86_64/"
  ],
  "search": "xfce"
}`

- add a repository that has a comps.xml file (e.g. https://dl.fedoraproject.org/pub/epel/8/Everything/x86_64/) and make a request to `repositories/{uuid}/package_groups`, should list package groups associated with that repo
- make a request to `package_groups/names` with a valid repo URL or repo UUID and a search term, should return the package groups with that search term in the name
** for example, the epel 8 and epel 9 repos have some package groups (one is `xfce`) with different package lists. if you add both of these repos and include both repos in the request, the search API should return the full list of packages from both repos like this:
`[
    {
        "package_group_name": "Xfce",
        "description": "A lightweight desktop environment that works well on low end machines.",
        "package_list": [
            "xfce4-screensaver",
            "tumbler",
            "NetworkManager-gnome",
            "openssh-askpass",
            "gdm",
            "xfce4-terminal",
            "thunar-volman",
            "xfce4-panel",
            "xfce4-appfinder",
            "network-manager-applet",
            "xfwm4",
            "xfce4-settings",
            "xfce4-pulseaudio-plugin",
            "pinentry-gtk",
            "mousepad",
            "xfdesktop",
            "xfconf",
            "Thunar",
            "xfce-polkit",
            "thunar-archive-plugin",
            "xfce4-session",
            "pinentry-gnome3",
            "xfce4-power-manager"
        ]
    }
]`

## Checklist

- [x] Tested with snapshotting feature disabled and pulp server URL not configured if appropriate

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-2443

File: pkg/api/repository_package_groups.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+type SearchPackageGroupRequest struct {
+	URLs   []string `json:"urls,omitempty"`  // URLs of repositories to search
+	UUIDs  []string `json:"uuids,omitempty"` // List of RepositoryConfig UUIDs to search
+	Search string   `json:"search"`          // Search string to search package group names
+	Limit  *int     `json:"limit,omitempty"` // Maximum number of records to return for the search
+}
+
+const SearchPackageGroupRequestLimitDefault int = 100
+const SearchPackageGroupRequestLimitMaximum int = 500
```
Comment (rverdile) [inline pkg/api/repository_package_groups.go:33]: could potentially be shared

File: pkg/dao/package_groups.go [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+}
+
+func GetPackageGroupDao(db *gorm.DB) PackageGroupDao {
+	// Return DAO instance
+	return packageGroupDaoImpl{
+		db: db,
+	}
+}
+
+func (r packageGroupDaoImpl) isOwnedRepository(orgID string, repositoryConfigUUID string) (bool, error) {
```
Comment (rverdile) [inline pkg/dao/package_groups.go:29]: could potentially be shared

File: pkg/dao/package_groups.go [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+	}
+
+	// FIXME 103 Once the URL stored in the database does not allow
+	//           "/" tail characters, this could be removed
+	urls := make([]string, len(request.URLs)*2)
+	for i, url := range request.URLs {
+		urls[i*2] = url
+		urls[i*2+1] = url + "/"
+	}
+	uuids := request.UUIDs
```
Comment (rverdile) [inline pkg/dao/package_groups.go:168]: could potentially be shared

File: pkg/dao/package_groups.go [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+	}
+	if len(request.URLs) == 0 && len(request.UUIDs) == 0 {
+		return nil, fmt.Errorf("must contain at least 1 URL or 1 UUID")
+	}
+	if request.Limit == nil {
+		request.Limit = pointy.Int(api.SearchPackageGroupRequestLimitDefault)
+	}
+	if *request.Limit > api.SearchPackageGroupRequestLimitMaximum {
+		request.Limit = pointy.Int(api.SearchPackageGroupRequestLimitMaximum)
+	}
```
Comment (rverdile) [inline pkg/dao/package_groups.go:159]: could potentially be shared

File: pkg/handler/repository_package_groups.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+	apiResponse, err := rh.Dao.PackageGroup.Search(orgId, dataInput)
+	if err != nil {
+		return ce.NewErrorResponse(http.StatusInternalServerError, "Error searching package groups", err.Error())
+	}
+
+	return c.JSON(200, apiResponse)
+}
+
+func (rh *RepositoryPackageGroupHandler) searchPackageGroupPreprocessInput(input *api.SearchPackageGroupRequest) {
```
Comment (rverdile) [inline pkg/handler/repository_package_groups.go:58]: could be potentially be shared

**Review by rverdile — COMMENTED** [review]
I think there may be some places where we can create shared functions or structs between rpm, package groups, and environments. And in the future, snapshot rpms, package groups, and environments. All 6 of these things have a similar flow, so it probably makes sense to try and de-duplicate where possible.

I left comments where I think might be potential places to create shared code. With that said, it probably make sense to leave it as-is here, and make changes to the environments PR instead. This way you can see all 3 at the same time and easily spot any duplication.

File: pkg/dao/package_groups.go [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+		return
+	}
+	out.UUID = in.Base.UUID
+	out.ID = in.ID
+	out.Name = in.Name
+	out.Description = in.Description
+	out.PackageList = in.PackageList
+}
+
+func (r packageGroupDaoImpl) Search(orgID string, request api.SearchPackageGroupRequest) ([]api.SearchPackageGroupResponse, error) {
```
Comment (rverdile) [inline pkg/dao/package_groups.go:146]: For some reason when I search EPEL 9 or EPEL 7, I get `[]`. It looks like it introspected correctly and I see its package groups in the database, but I don't see anything in the search.

The search for EPEL 8 works just fine.

```
POST http://localhost:8000/api/content-sources/v1.0/package_groups/names
x-Rh-Identity: eyJpZGVudGl0eSI6eyJ0eXBlIjoiVXNlciIsInVzZXIiOnsidXNlcm5hbWUiOiJzbmFwVXNlciJ9LCJhY2NvdW50X251bWJlciI6IjEwMDgiLCJpbnRlcm5hbCI6eyJvcmdfaWQiOiIxMDA4In19fQo=
Content-Type: application/json

{
  "urls": ["https://dl.fedoraproject.org/pub/epel/9/Everything/x86_64/"],
  "search": "xfce"
}
```

Comment (rverdile) [reply pkg/dao/package_groups.go:146]: Do you see that too? If so, there might be something up with the search query. Trying to debug.

Comment (xbhouse) [reply pkg/dao/package_groups.go:146]: that's weird. i tried the search api with these payloads and receive the correct data:

`{
  "search": "xfce",
  "urls": [
    "https://dl.fedoraproject.org/pub/epel/9/Everything/x86_64/"
  ]
}`

`{
  "search": "xfce",
  "urls": [
    "https://dl.fedoraproject.org/pub/epel/8/Everything/x86_64/"
  ]
}`

`{
  "search": "xfce",
  "urls": [
    "https://dl.fedoraproject.org/pub/epel/7/x86_64/"
  ]
}`

which repo urls and what did your request look like? i see that i get `[]` if introspection failed though

**Comment by xbhouse** [comment]
> I think there may be some places where we can create shared functions or structs between rpm, package groups, and environments. And in the future, snapshot rpms, package groups, and environments. All 6 of these things have a similar flow, so it probably makes sense to try and de-duplicate where possible.
> 
> I left comments where I think might be potential places to create shared code. With that said, it probably make sense to leave it as-is here, and make changes to the environments PR instead. This way you can see all 3 at the same time and easily spot any duplication.

that makes sense! i will make those changes in the environments PR

File: pkg/dao/package_groups.go (hunk 6 above)
Comment (rverdile) [reply pkg/dao/package_groups.go:146]: Looks like I was just using the wrong org ID. No issue here!

File: pkg/dao/package_groups_test.go
Comment (rverdile) [inline pkg/dao/package_groups_test.go:1]: Correct me if I missed it, but can you add a test to check for the package list concatentation/de-duplication when searching for something like Xfce?

**Review by rverdile — COMMENTED** [review]
This looks good, just needs a rebase and potentially that one test

File: pkg/dao/package_groups_test.go
Comment (xbhouse) [reply pkg/dao/package_groups_test.go:1]: added in the test case for dedup / concat. i realized i wasn't checking for the expected packagelist in the search at all so fixed that too :)

**Review by rverdile — APPROVED** [review]
nice job!! :tada: :tada: :tada:

**Comment by swadeley** [comment]
Hi @xbhouse 

Thank you

`In [7]: app.content_sources.rest_client.repositories_api.search_package_group({"search": "Xfce", "urls": ["https://dl.fedoraproject.org/pub/epel/8
   ...: /Everything/x86_64/"]})
Out[7]: 
[{'description': 'A lightweight desktop environment that works well on low end '
                 'machines.',
  'package_group_name': 'Xfce',
  'package_list': ['xfce4-screensaver',
                   'tumbler',
                   'NetworkManager-gnome',
                   'openssh-askpass',
                   'gdm',
                   'xfce4-terminal',
                   'thunar-volman',
                   'xfce4-panel',
                   'xfce4-appfinder',
                   'xfwm4',
                   'xfce4-settings',
                   'xfce4-pulseaudio-plugin',
                   'pinentry-gtk',
                   'mousepad',
                   'xfdesktop',
                   'xfconf',
                   'Thunar',
                   'xfce-polkit',
                   'thunar-archive-plugin',
                   'xfce4-session',
                   'xfce4-power-manager']}]
`

**Pull request merged**